src/ReactBoilerplate/gulpfile.js

```diff
@@ -1,54 +1,251 @@
-  });
-});
+// Subtasks
+const getWebpackConfigLoader = (configName) => {
+  const loader = (callback) => {
+    if (!webpackConfig[configName + 'Compiler']) {
+      webpackConfig[configName + 'Compiler'] = webpack(webpackConfig.config[configName]);
+    }
+    callback();
+  };
+  loader.displayName = 'load-webpack-config';
+  return loader;
+};
 
-// server
-gulp.task('watch-server', ['build-server'], function(){
-  gulp.watch(["Scripts/**/*"], ["build-server"]);
-})
-gulp.task('build-server-compiler', function() {
-  if(!webpackConfig.serverCompiler)
-    webpackConfig.serverCompiler = webpack(webpackConfig.config.server);
+const getBundleBuilder = (configName) => {
+  const builder = (callback) => {
+    webpackConfig[configName + 'Compiler'].run((err, stats) => {
+      if (err) throw new PluginError('build-bundle', err);
+      log(configName + ' Bundle ' + stats.toString({
+        colors: true,
+        chunks: false
+      }));
+      callback();
+    });
+  };
+  builder.displayName = 'build-bundle';
+  return builder;
+};
+
+const getBundleWatcher = (taskToTrigger) => {
+  const watcher = (callback) => {
+    gulp.watch(['Scripts/**/*'], gulp.series(taskToTrigger));
+    callback();
+  };
+  watcher.displayName = 'watch-bundle';
+  return watcher;
+};
+
+// Set production build flag
+gulp.task('prod', (callback) => {
+  global.isProduction = true;
+  log('isProduction=' + global.isProduction);
+  setWebpackConfig();
+  callback();
 });
-gulp.task('build-server', ['build-server-compiler'], function(cb) {
-  webpackConfig.serverCompiler.run(function(err, stats) {
-    if(err) throw new gutil.PluginError("build-server", err);
-    gutil.log("[build-server]", stats.toString({
-      colors: true,
-      chunks: false
+
+// Common Bundle tasks
+gulp.task('create-pack-dir', () =>
+  Promise.resolve()
+    .then(() => {
+      if (!fs.existsSync(packDir)) {
+        fs.mkdirSync(packDir);
+      }
     }));
-    cb();
-  });
+
+gulp.task('clean-bundles', () =>
+  Promise.resolve()
+    .then(() => del([packDir + '*'], { dot: true })));
+
+// Client bundles
+gulp.task(
+  'build-client-compiler',
+  gulp.series('create-pack-dir', getWebpackConfigLoader('client'))
+);
+
+gulp.task(
+  'build-client-bundle',
+  gulp.series('build-client-compiler', getBundleBuilder('client'))
+);
+
+gulp.task('build-dummy-style-bundle', (callback) => {
+  fs.writeFile(packDir + 'styles.css', '', callback);
 });
+
+gulp.task(
+  'watch-client-bundle',
+  gulp.series('build-client-bundle', getBundleWatcher('build-client-bundle'))
+);
+
+// Server Bundles
+gulp.task(
+  'build-server-compiler',
+  gulp.series('create-pack-dir', getWebpackConfigLoader('server'))
+);
+
+gulp.task(
+  'build-server-bundle',
+  gulp.series('build-server-compiler', getBundleBuilder('server'))
+);
+
+gulp.task(
+  'watch-server-bundle',
+  gulp.series('build-server-bundle', getBundleWatcher('build-server-bundle'))
+);
+
+// Server .NET Core
+gulp.task('restore-dotnet-packages', () =>
```

Comment:
@neodescis, I don't think the gulpfile should be responsible for the dotnet stuff. We could update the csproject to hook up to the Build/Clean targets to run related gulp tasks, but the over-arching build process of dotnet should be left up to the implementation (wether they use Cake, dotnet-script, make, etc).

Can you remove this?

Reply:
@pauldotknopf, I can take it out if you like; I like gulp for orchestrating, so this is how I've been running things for development. Perhaps more importantly though, if I remove this, then the webpack dev middleware and HMR go with it, since they are tied into browsersync and its proxy to the dotnet server. You sure you want me to take it out? Up to you.